new ELIP: Confidential Transaction descriptors

apoelstra opened the pull request:
This ELIP describes CT descriptors, as needed to attach blinding keys to existing descriptors and therefore create confidential addresses.

It does **not** cover:
* Ordinary Elements descriptors, e.g. `eltr`; these need a separate standard preceding this one
* Pegins or pegouts
* Anything related to PAK keys

jgriffiths commented on elip-ct-descriptors.mediawiki:
```diff
@@ -0,0 +1,96 @@
+* `hash_to_sk(<key1>, <key2>, ...)` which indicates that the blinding key is produced by a sorted hash of the listed key(s), which will be one or more public keys; or
+* `<key>` which is an ordinary (extended) public key which will be used as a blinding key
+
+We note that even if the descriptor includes secret keys, as it will when used for wallet backups, the keys included in the `hash_to_sk` combinator will be public keys.
```
This could be clearer IMO, below it states `'''hash_to_sk''' takes a list of public or private keys`. I take it the meaning is that the hashed value is computed on the public keys, and its the public keys that are sorted before being hashed. Would be good to be explicit/consistent between here and the `hash_to_sk` description below.

apoelstra replied:
Yes, the meaning is that the hash is computed using the public keys.

It would be quite difficult for implementors to force public keys in the actual descriptor even when they have private keys elsewhere, since they (probably) have only a single key type. So saying that they *must* be public keys is wrong. What I mean to say is that they will be coerced into secp256k1 public keys, in the same way that they would be if they were included in a script, and then hashed.

real-or-random replied:
> Would be good to be explicit/consistent between here and the `hash_to_sk` description below.

Agree. Do you have a suggestion?

jgriffiths replied:
Perhaps something like:
```
Note that if the descriptor includes secret keys, as it will when used for wallet backups, the `hash_to_sk` combinator will convert them to public keys internally before sorting and computing its hash result.
```
?

jgriffiths commented on elip-ct-descriptors.mediawiki:
```diff
@@ -0,0 +1,96 @@
+    * every public key, in 33-byte compressed encoding, concatenated in lexicographic order
+The output of this hash, when interpreted as an integer (most significant byte first) modulo the group order, is the blinding ''secret'' key.
+* '''bare key''', which simply encodes a single key in the same way as the keys in `ordinary descriptor` are encoded.
+This may be a private or public key; it is used as the blinding key for the address. 
```
```suggestion
This may be a private or public key; its public key is used as the blinding key for the address. 
```

real-or-random replied:
Hm @apoelstra wrote that part but as I understand it, it should be possible to specify a private blinding key, too?

jgriffiths replied:
I'm trying to capture here that the confidential address only includes the public blinding key. If a private key is given, the address generated will include the public key corresponding to the private key.

jgriffiths reviewed (commented):
Notes:

1. I strongly prefer to use 'private key' rather than 'secret key' for EC key pair terminology, this matches other BIPs better, and as noted here here there may be confusion with 'secret' public keys. My suggested changes use 'secret' just because thats the minimum review delta, but I think it should be globally changed so the only reference to secrecy is the notion of confusion around secret public keys.

2. Similarly does the naming of `hash_to_sk` conflict with any current or proposed descriptor/miniscript operator that produces or takes private keys? I feel like `hash_to_scalar` or something other than `sk` would be a better name.

3. There is a negligible but non-zero chance that the `hash_to_sk` construction will produce an invalid key, a resolution or suggested action should this occur should be listed for completeness, as one done with bip32 despite it being for all intents and purposes cryptographically unreachable.

apoelstra commented on elip-ct-descriptors.mediawiki:
```diff
@@ -0,0 +1,96 @@
+
+`<key>` is one of the following three options
+* '''slip77''', encoded as `slip77(<64 hex characters>)`, whose semantics are that the 64 hex characters are interpreted as the 32-byte `master_blinding_key` in SLIP77; the `scriptPubKey` for SLIP77 is computed as normal from the ordinary descriptor.
+For both private and public descriptors, these 32 bytes are encoded identically.
```
Yep, thanks! That is a useful clarification especially since for the non-SLIP77 cases, there is a meaningful distinction between private and public keys.

apoelstra replied:
Changed to `These 32 bytes will be the same for both private and public descriptors; they have no "public" equivalent.'

apoelstra commented:
Thanks @jgriffiths for your detailed review! I will accept all your suggestiotns, except to change `secret` to `private` even in the places that you used "secret".

Regarding the name `hash_to_sk`, what if I used `hash_to_private`? I don't want to use `scalar` because I'd like to indicate to the user that the output of this function is private key data, and that anyone who is able to compute it (i.e. anyone with access to the descriptor) can unblind outputs.

jgriffiths commented:
Also missing I think are a reference implementations section listing the elements-miniscript impl (until there are more impls), and a set of test vectors.

For test vectors for `hash_to_private` I think a mix of public and private keys out of public-key-order would be good, my preference is the input string and expected output private key encoded in WIF. Perhaps we want to specify that all `hash_to_private` keys must belong to the same network and have an invalid test vector which mixes them. Also an invalid vector with an input that is not a public/private key.

sanket1729 commented on elip-ct-descriptors.mediawiki:
```diff
@@ -0,0 +1,96 @@
+Here `<ordinary descriptor>` refers to any existing descriptor, e.g. `wsh(...)` or `tr(...)` and `<key>` refers to a special construction which must be one of:
+* `slip77(<master secret>)` which indicates that blinding keys for these addresses are derived via SLIP77; or
+* `hash_to_sk(<key1>, <key2>, ...)` which indicates that the blinding key is produced by a sorted hash of the listed key(s), which will be one or more public keys; or
+* `<key>` which is an ordinary (extended) public key which will be used as a blinding key
```
I think we should rename `<key>` to `BLINDING_KEY` (note the caps). `KEY` expressions are well-defined in descriptors [spec.](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#reference). In the descriptor spec, all the expressions are using the `CAPS` so it would be good to be consistent.

We can then write:
```md
`BLINDING_KEY` expressions are either:
- slip77(<`64-byte-hex`>)
- hash_to_sk(<`KEY_1`>, <`KEY_2`>, <`KEY_3`>.... <`KEY_n`>)
- `KEY`
```
while referring to the key as defined in the spec.

apoelstra replied:
My current reference logic just silently compresses uncompressed keys. Probably we should forbid them instead.

jgriffiths replied:
If we are voting, I vote for disallowing please.

sanket1729 reviewed (commented):
Looks good. Suggested a minor rewrite to be consistent how it's written in descriptors.md

real-or-random commented:
>     1. I strongly prefer to use 'private key' rather than 'secret key' for EC key pair terminology, this matches other BIPs better, and as noted here here there may be confusion with 'secret' public keys. My suggested changes use 'secret' just because thats the minimum review delta, but I think it should be globally changed so the only reference to secrecy is the notion of confusion around secret public keys.

I don't agree but I don't have a strong opinion here. I think "secret" and "private" are interchangeable and both are good. libsecp256k1 uses "secret", and "secret" has the tiny advantage that you can write sk and pk. 

>     2. Similarly does the naming of `hash_to_sk` conflict with any current or proposed descriptor/miniscript operator that produces or takes private keys? I feel like `hash_to_scalar` or something other than `sk` would be a better name.

I think it the name should make clear that the output is secret (or private). Or do you see a change that this is reused in a context where the result is not secret?

> 
>     3. There is a negligible but non-zero chance that the `hash_to_sk` construction will produce an invalid key, a resolution or suggested action should this occur should be listed for completeness, as done with bip32 for example, despite it being for all intents and purposes cryptographically unreachable.

Actually, what's a valid secret blinding key? The same as a signing key, so 0 is invalid? Or is 0 actually allowed? 

Assuming that 0 is the only scalar which is an invalid secret blinding key, this happens when the hash is 0 or the group order, so the probability of this happening is 2/2^-256 = 2^-255 when SHA256 is modeled as a RO.  I'm not strictly against specifying what to do in these cases but any discussion about this is a waste of time, as this simply won't happen in practice. A good choice is aborting or incrementing by 1.

real-or-random commented:
> Assuming that 0 is the only scalar which is an invalid secret blinding key, this happens when the hash is 0 or the group order, so the probability of this happening is 2/2^-256 = 2^-255 when SHA256 is modeled as a RO. I'm not strictly against specifying what to do in these cases but discussion about this tend to be a waste of time, as this simply won't happen in practice. A good choice is aborting or incrementing by 1.

Actually the draft implementation does not match the spec, which says take the hash mod the group order. But we can also change the spec to simply forbid every hash which is too large. None of this really matters in the end.

jgriffiths commented:
I missed the hash being taken mod n, if thats the case then then no further action is needed.

apoelstra commented:
* changed "secret key" to "private key" everywhere on account of Jon caring the most
* added a "if this hash is zero the resultineg descriptor is invalid" sentence
* rename `hash_to_sk` to `hash_to_private`
* addressed all other comments

At this point the remaining TODOs are 
* Update reference impl to use hash_to_private (and address comments on that PR)
* Generate test vectors from the reference impl

apoelstra commented:
I don't know what's wrong with github. It shows new commits on my branch but they aren't appearing on this PR.

apoelstra commented:
I have a couple questions:
* when serializing descriptors as strings, should the keys be sorted at all? should we require they be sorted when parsing?
* should two descriptors compare equal if they differ only in the order of `hash_to_private` keys?

Bear in mind that the keys in the descriptor may be xpubs, potentially with wildcards, and therefore will **not** sort the same way as the EC keys that get hashed.

apoelstra commented:
Ok @jgriffiths, makes sense. I won't try enforcing any sort order (or making any internal operations sort-order-independent).

For what it's worth, it's not that we "missed the boat", but that we simply could not come up with a canonical representation. `'` vs `h`, sort order, etc., is easy enough, but deciding when xpubs of different length are equivalent was not. (You can fingerprint the final derived path, but not when wildcards are present, and we technically support hardened wildcards and post-wildcard hardened steps which makes "derive at some fixed index" impossible, and so on ... and this is not even touching issues with taproot/partial descriptors that we knew we'd eventually have to deal with).

apoelstra commented:
Neat, apparently Github **does** allow me to take actions that will permanently auto-close the PR without warning, it just prevents me from doing anything useful.

Moving to #2, sorry for the confusion.

jgriffiths commented:
> For what it's worth, it's not that we "missed the boat", but that we simply could not come up with a canonical representation.

Yes, I think you can have a canonical representation only for a given collection of conditions excluding the actual keys. That allows you to say things like 'these are bip44 compatible' or 'this is a BOLT-standard spending path'  but not compare two expressions where say one is for public consumption and the other is used by the signer.
